### PR TITLE
Stabilize upgrade codes for host msi

### PR DESCRIPTION
**Pull request by NikolaMilosavljevic**
Stable upgrade codes and dependency key to allow proper msi upgrade experience.

We are using values from latest official release (6.0.0).

### CODE REVIEW

**Comment by joeloff** [comment]
We should probably look at the dependency provider key and figure out what to do. Below is from RC2 and RTM. When upgrading, we need to ensure stable key so that chainer dependencies can be maintained, otherwise we risk losing the ref count against the MSI itself.

image

**Comment by NikolaMilosavljevic** [comment]
> We should probably look at the dependency provider key and figure out what to do. Below is from RC2 and RTM. When upgrading, we need to ensure stable key so that chainer dependencies can be maintained, otherwise we risk losing the ref count against the MSI itself.
> 
> image

Thanks. I'm going to see what the best option is to override the auto-generated values.

**Comment by joeloff** [comment]
> > We should probably look at the dependency provider key and figure out what to do. Below is from RC2 and RTM. When upgrading, we need to ensure stable key so that chainer dependencies can be maintained, otherwise we risk losing the ref count against the MSI itself.
> > image
> 
> Thanks. I'm going to see what the best option is to override the auto-generated values.
 Something like `Dotnet_CLI_SharedHost,$(MajorVersion).$(MinorVersion),$(Platform)` should work (or whatever the relevant properties are). The tricky bit is whether we decide to leave the older versions or if we allow them to upgrade, allow the ref count to become inaccurate as part of the transition.

**Comment by NikolaMilosavljevic** [comment]
> > > We should probably look at the dependency provider key and figure out what to do. Below is from RC2 and RTM. When upgrading, we need to ensure stable key so that chainer dependencies can be maintained, otherwise we risk losing the ref count against the MSI itself.
> > > image
> > 
> > 
> > Thanks. I'm going to see what the best option is to override the auto-generated values.
> > Something like `Dotnet_CLI_SharedHost,$(MajorVersion).$(MinorVersion),$(Platform)` should work (or whatever the relevant properties are). The tricky bit is whether we decide to leave the older versions or if we allow them to upgrade, allow the ref count to become inaccurate as part of the transition.

I was thinking of allowing override of ProviderKey value, which works in my tests and is in my new Arcade PR: https://github.com/dotnet/arcade/pull/8181

**Comment by NikolaMilosavljevic** [comment]
I've pushed the new change to stabilize ProviderKey value by overriding the Candle variable.

**Comment by NikolaMilosavljevic** [comment]
> I've pushed the new change to stabilize ProviderKey value by overriding the Candle variable.

Will also need to update a reference to Arcade's Installers package once Arcade PR is approved and we get a build: https://github.com/dotnet/arcade/pull/8181

**Comment by NikolaMilosavljevic** [comment]
I've updated the Arcade package (Microsoft.DotNet.Build.Tasks.Installers). This PR can now be merged, once I get the approving review and it finishes building.

**Review by joeloff — APPROVED** [review]

File: src/installer/pkg/sfx/installers/dotnet-host.proj [hunk 1]
```diff
@@ -21,12 +21,20 @@
     <RpmScriptsDirectory>$(MSBuildThisFileDirectory)rpm_scripts/host</RpmScriptsDirectory>
     <RpmAfterInstallScript>$(RpmScriptsDirectory)/after_install.sh</RpmAfterInstallScript>
     <RpmAfterRemoveScript>$(RpmScriptsDirectory)/after_remove.sh</RpmAfterRemoveScript>
+    <!-- Stabilizes upgrade codes, using values from 6.0.0 release - do not change -->
```
Comment (ericstj) [inline src/installer/pkg/sfx/installers/dotnet-host.proj:24]: Make sure you figure out how to backport equivalent logic to 7.0.  Hopefully without using GUIDs we have to regenerate every release.  Either something stable, or something generated with the right set of inputs.

File: src/installer/pkg/sfx/installers/dotnet-host.proj [hunk 2]
```diff
@@ -21,12 +21,20 @@
...
+    <UpgradeCode Condition="'$(TargetArchitecture)' == 'x64'">{97C9F490-5379-6902-7B80-D2723CA59D95}</UpgradeCode>
+    <UpgradeCode Condition="'$(TargetArchitecture)' == 'x86'">{181630A5-53EA-4D70-E563-06F64946DE38}</UpgradeCode>
   </PropertyGroup>
 
   <ItemGroup>
     <WixSrcFile Include="host.wxs" />
     <WixExtraComponentGroupRefId Include="InstallSharedHostandDetectionKeys" />
     <CandleVariables Include="ExtraPropertyRefIds" Value="ProductCPU;RTM_ProductVersion" />
+    <!-- Stabilizes provider key, using values from 6.0.0 release - do not change -->
+    <CandleVariables Condition="'$(TargetArchitecture)' == 'arm64'" Include="DependencyKey" Value="Dotnet_CLI_SharedHost_48.3.31210_arm64" />
```
Comment (ericstj) [inline src/installer/pkg/sfx/installers/dotnet-host.proj:35]: No clue what that version string means, but I trust it's correct?

**Review by ericstj — APPROVED** [review]

File: src/installer/pkg/sfx/installers/dotnet-host.proj (hunk 1 above)
Comment (NikolaMilosavljevic) [reply src/installer/pkg/sfx/installers/dotnet-host.proj:24]: Yes, 7.0 will use stable inputs as upgrade code seed. Major.minor, arch, and generic host name.

**Comment by safern** [comment]
@danmoseley is this approved to merge for this servicing?

**Comment by NikolaMilosavljevic** [comment]
This test failure seems to be a known issue, that is breaking other servicing PRs: https://github.com/dotnet/runtime/issues/61636

**Comment by NikolaMilosavljevic** [comment]
> @danmoseley is this approved to merge for this servicing?

This is all part of the same setup work, for 6.0, 5.0 and 3.1. Here are few other examples:
https://github.com/dotnet/runtime/pull/61619
https://github.com/dotnet/arcade/pull/8181
https://github.com/dotnet/arcade/pull/8174

**Comment by safern** [comment]
Test failure is: #61636

**Pull request merged**